src/macros.rs

```diff
@@ -18,10 +18,83 @@ macro_rules! impl_pretty_debug {
         impl ::core::fmt::Debug for $thing {
             fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
                 write!(f, "{}(", stringify!($thing))?;
-                for i in self[..].iter().cloned() {
+                for i in &self[..] {
                     write!(f, "{:02x}", i)?;
                 }
-                write!(f, ")")
+                f.write_str(")")
+            }
+        }
+     }
+}
+
+macro_rules! impl_safe_debug {
+    ($thing:ident) => {
+        #[cfg(feature = "alloc")]
+        use alloc::string::String;
+
+        #[cfg(feature = "bitcoin_hashes")]
```

Comment:
If all we want is differentiating different secret keys based on a hash, is it worth it going through the effort of using `bitcoin_hashes` for this? Under what circumstances is this better than just always using `hash_map::DefaultHasher`?

Reply:
I think that's a bad idea. We should use a preimage-resistant hash to make sure you can't recover any key bits from the hash. The std lib does not give it many guarantees about `DefaultHasher`: https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#22 The algorithm is currently SipHash, which is not a hash function but a PRF with a seed generated at run time, so the output changes in different runs of the program.

Reply:
Right that makes sense. It does mean though we can only provide a `Debug` impl if we have the `bitcoin_hashes` feature enabled? Or would you print something like `[redacted]` in the other case?

Reply:
Maybe we should print: `[redacted but if you want to see a hash enabled the "bitcoin_hashes" feature of 'rust-bitcoin']`.

Reply:
Or `[redacted, see docs]` for brevity.

Reply:
Yeah, I wasn't entirely serious with the 2nd suggestion :)